### ISSUE TITLE
0.9.2 Regression: Okta login fails when Duo MFA is required

### ISSUE THREAD

**Issue by jgiles** [hashicorp/vault#3872]
**Environment:**

* Vault Version: Vault v0.9.2 ('3bf8733cd69bb0ac14da9aaa6135bcb7f710cc5f')
* Operating System/Architecture: Docker on OSX

**Vault Config File:**

Default dev server config

**Startup Log Output:**

```
Couldn't start vault with IPC_LOCK. Disabling IPC_LOCK, please use --privileged or --cap-add IPC_LOCK
==> Vault server configuration:

                     Cgo: disabled
         Cluster Address: https://0.0.0.0:8201
              Listener 1: tcp (addr: "0.0.0.0:8200", cluster address: "0.0.0.0:8201", tls: "disabled")
               Log Level: debug
                   Mlock: supported: true, enabled: false
        Redirect Address: http://0.0.0.0:8200
                 Storage: inmem
                 Version: Vault v0.9.2
             Version Sha: 3bf8733cd69bb0ac14da9aaa6135bcb7f710cc5f

WARNING! dev mode is enabled! In this mode, Vault runs entirely in-memory
and starts unsealed with a single unseal key. The root token is already
authenticated to the CLI, so you can immediately begin using Vault.

You may need to set the following environment variable:

    $ export VAULT_ADDR='http://0.0.0.0:8200'

The unseal key and root token are displayed below in case you want to
seal/unseal the Vault or re-authenticate.

Unseal Key: 05ES4Mqk5G6lE0XZW6rWYoqorKx/o6vxmz/E/yGB08E=
2018/01/30 21:59:12.030404 [INFO ] core: security barrier not initialized
2018/01/30 21:59:12.030527 [INFO ] core: security barrier initialized: shares=1 threshold=1
2018/01/30 21:59:12.030780 [DEBUG] core: cluster name set: name=vault-cluster-47da089b
2018/01/30 21:59:12.030790 [DEBUG] core: cluster ID set: id=4d78d23f-b62e-47c5-cbd4-8fd155db4d29
2018/01/30 21:59:12.030829 [INFO ] core: post-unseal setup starting
2018/01/30 21:59:12.044912 [INFO ] core: loaded wrapping token key
2018/01/30 21:59:12.044928 [INFO ] core: successfully setup plugin catalog: plugin-directory=
2018/01/30 21:59:12.049823 [INFO ] core: successfully mounted backend: type=kv path=secret/
2018/01/30 21:59:12.049843 [INFO ] core: successfully mounted backend: type=cubbyhole path=cubbyhole/
2018/01/30 21:59:12.049956 [INFO ] core: successfully mounted backend: type=system path=sys/
Root Token: test-root-token

Development mode should NOT be used in production installations!

==> Vault server started! Log data will stream in below:

2018/01/30 21:59:12.050098 [INFO ] core: successfully mounted backend: type=identity path=identity/
2018/01/30 21:59:12.051321 [INFO ] expiration: restoring leases
2018/01/30 21:59:12.051424 [INFO ] rollback: starting rollback manager
2018/01/30 21:59:12.051491 [DEBUG] expiration: collecting leases
2018/01/30 21:59:12.051765 [DEBUG] expiration: leases collected: num_existing=0
2018/01/30 21:59:12.052089 [INFO ] expiration: lease restore complete
2018/01/30 21:59:12.052335 [DEBUG] identity: loading entities
2018/01/30 21:59:12.052342 [DEBUG] identity: entities collected: num_existing=0
2018/01/30 21:59:12.052393 [INFO ] identity: entities restored
2018/01/30 21:59:12.052396 [DEBUG] identity loading groups
2018/01/30 21:59:12.052402 [DEBUG] identity: groups collected: num_existing=0
2018/01/30 21:59:12.052405 [INFO ] identity: groups restored
2018/01/30 21:59:12.052412 [INFO ] core: post-unseal setup complete
2018/01/30 21:59:12.052721 [INFO ] core: root token generated
2018/01/30 21:59:12.052728 [INFO ] core: pre-seal teardown starting
2018/01/30 21:59:12.052732 [INFO ] core: cluster listeners not running
2018/01/30 21:59:12.052755 [DEBUG] expiration: stop triggered
2018/01/30 21:59:12.052762 [DEBUG] expiration: finished stopping
2018/01/30 21:59:12.053634 [INFO ] rollback: stopping rollback manager
2018/01/30 21:59:12.053808 [INFO ] core: pre-seal teardown complete
2018/01/30 21:59:12.054033 [INFO ] core: vault is unsealed
2018/01/30 21:59:12.054065 [INFO ] core: post-unseal setup starting
2018/01/30 21:59:12.054283 [INFO ] core: loaded wrapping token key
2018/01/30 21:59:12.054290 [INFO ] core: successfully setup plugin catalog: plugin-directory=
2018/01/30 21:59:12.054619 [INFO ] core: successfully mounted backend: type=kv path=secret/
2018/01/30 21:59:12.054851 [INFO ] core: successfully mounted backend: type=system path=sys/
2018/01/30 21:59:12.055429 [INFO ] core: successfully mounted backend: type=identity path=identity/
2018/01/30 21:59:12.055954 [INFO ] core: successfully mounted backend: type=cubbyhole path=cubbyhole/
2018/01/30 21:59:12.057243 [INFO ] expiration: restoring leases
2018/01/30 21:59:12.057282 [DEBUG] identity: loading entities
2018/01/30 21:59:12.057289 [DEBUG] identity: entities collected: num_existing=0
2018/01/30 21:59:12.057317 [INFO ] rollback: starting rollback manager
2018/01/30 21:59:12.057344 [DEBUG] expiration: collecting leases
2018/01/30 21:59:12.057362 [DEBUG] expiration: leases collected: num_existing=0
2018/01/30 21:59:12.057454 [INFO ] expiration: lease restore complete
2018/01/30 21:59:12.057476 [INFO ] identity: entities restored
2018/01/30 21:59:12.057481 [DEBUG] identity loading groups
2018/01/30 21:59:12.057492 [DEBUG] identity: groups collected: num_existing=0
2018/01/30 21:59:12.057495 [INFO ] identity: groups restored
2018/01/30 21:59:12.057504 [INFO ] core: post-unseal setup complete
2018/01/30 21:59:45.215596 [INFO ] core: enabled credential backend: path=okta/ type=okta
2018/01/30 22:00:13.957018 [DEBUG] auth/okta: unhandled result status: status=MFA_REQUIRED
2018/01/30 22:12:11.700812 [DEBUG] auth/okta: unhandled result status: status=MFA_REQUIRED
2018/01/30 22:19:45.011411 [INFO ] expiration: revoked lease: lease_id=auth/token/create/a6e8118490127a427df48a308157a52074ab6340
2018/01/30 22:19:45.210615 [INFO ] expiration: revoked lease: lease_id=auth/token/create/a2d9e856b5436cddf6a7d8be670f5732511de809
```

**Expected Behavior:**

I can log in with Okta and Duo MFA, with a Duo Push. Confirmed that this works in v0.9.1:
```
$ vault status
Key             Value
---             -----
Seal Type       shamir
Sealed          false
Total Shares    1
Threshold       1
Version         0.9.1
Cluster Name    vault-cluster-269604c3
Cluster ID      0d04e920-61cf-538e-a77d-f81217ef2c35
HA Enabled      false

$ vault login -method=okta username=jgiles
Password (will be hidden):
Success! You are now authenticated. The token information displayed below
is already stored in the token helper. You do NOT need to run "vault login"
again. Future Vault requests will automatically use this token.

Key                    Value
---                    -----
token                  f0180af8-30f8-2a37-56a0-78e357aa8f0d
token_accessor         9ba0d922-a967-c0e8-9ec0-5235056b4aa6
token_duration         768h
token_renewable        true
token_policies         [default]
token_meta_policies    
token_meta_username    jgiles
```

**Actual Behavior:**

Authentication fails without sending a Duo Push.

```
$ vault status
Key             Value
---             -----
Seal Type       shamir
Sealed          false
Total Shares    1
Threshold       1
Version         0.9.2
Cluster Name    vault-cluster-47da089b
Cluster ID      4d78d23f-b62e-47c5-cbd4-8fd155db4d29
HA Enabled      false
$ vault login -method=okta username=jgiles
Password (will be hidden):
Error authenticating: Error making API request.

URL: PUT http://192.168.99.100:8200/v1/auth/okta/login/jgiles
Code: 400. Errors:

* okta authentication failed
```

With debug logs enabled, I see the following in the logs:
```
2018/01/30 22:00:13.957018 [DEBUG] auth/okta: unhandled result status: status=MFA_REQUIRED
```

I'm guessing it's related to the 0.9.2 change that denies blocked Okta accounts.

**Steps to Reproduce:**

See above.

**Important Factoids:**

The client version doesn't seem to matter, I tried with 0.9.1, 0.9.2, and 0.9.3 clients.

**References:**

**Comment by jefferai** [comment]
I believe @broamski may work up a patch for this.

**Comment by broamski** [comment]
Correct. I should have something by the end of the week.

Your auth commands look identical in both examples, what are you doing different to make the first auth successful?

**Comment by jgiles** [comment]
The commands and client are identical; the difference is that in the working case the running server is at v0.9.1, and in the broken case it is at v0.9.2

**Comment by broamski** [comment]
Ah, yes. Sorry. 

Out of curiosity, do you plan to use Okta’s MFA or Duo and what is your reasoning for the decision? I’m attempting to gather use cases other than my own.

**Comment by jgiles** [comment]
We've opted for Duo MFA because we're interested in a number of other features/services offered by Duo (endpoint certificates, network gateway, etc).

**Comment by shanewarner** [comment]
Verified that this is a breaking change in the enterprise version ...... We'll file a support ticket ....

**Comment by ccslamstack** [comment]
Okta Auth is broken period in Enterprise on 0.9.3 . Does Okta with Okta Push work at all in any of the product releases?   Push doesn't seem to be a factor here but okta is an issue.  This is halting us from going to production with Vault.  Please fix not only Okta Auth, but Enable Okta Push on okta auth for the mfa_method.

**Comment by broamski** [comment]
#3908  /cc @jgiles

**Comment by jefferai** [comment]
@shanewarner please note that this ticket is in reference to the deprecated MFA logic which we do support. Enterprise users should be using the MFA support discussed at https://www.vaultproject.io/docs/enterprise/mfa/index.html

**Comment by jefferai** [comment]
I meant, which we do *not* support.

**Comment by shanewarner** [comment]
I see. Well we are seeing the exact same issue and error message using the enterprise okta MFA functionality on this version and the previous version.

**Comment by shanewarner** [comment]
Looks like you guys changed the response handler and it doesn't know what to do with MFA_REQUIRED. My guess is that 0.9.1 silently ignored this and allowed MFA to work after auth succeeded.

+	case "LOCKED_OUT":
 +		if b.Logger().IsDebug() {
 +			b.Logger().Debug("auth/okta: user is locked out", "user", username)
 +		}
 +		return nil, logical.ErrorResponse("okta authentication failed"), nil, nil
 +
 +	case "PASSWORD_EXPIRED":
 +		if b.Logger().IsDebug() {
 +			b.Logger().Debug("auth/okta: password is expired", "user", username)
 +		}
 +		return nil, logical.ErrorResponse("okta authentication failed"), nil, nil
 +
 +	case "PASSWORD_WARN":
 +		oktaResponse.AddWarning("Your Okta password is in warning state and needs to be changed soon.")
 +
 +	case "SUCCESS":
 +		// Do nothing here
 +
 +	default:
 +		if b.Logger().IsDebug() {
 +			b.Logger().Debug("auth/okta: unhandled result status", "status", result.Status)
 +		}
 +		return nil, logical.ErrorResponse("okta authentication failed"), nil, nil
 +	}

**Comment by jefferai** [comment]
@shanewarner Pre 0.9.2/3, the status wasn't properly being checked, including MFA status. This made it easy for the Duo support introduced in 0.9.1 to work, but now that the status is being properly checked, the issue now is that when the status is returned as MFA_REQUIRED, how to handle that.

Ideally Okta would perform the MFA flow internally as part of the authentication API call, but it doesn't. So we're left with a couple of possibilities: rip out the Duo support and only support Okta MFA in the backend; bypass Okta MFA and only support the legacy Duo (which is clearly not the right option); or require both MFA checks to pass if both are configured. #3908 is tracking some work around this.

My inclination is to rip out the Duo support. Allowing it to be added was based on the belief that the backend was properly performing Okta flows, which we now know it wasn't, and it's only been in there for one (working) release. If someone wants to have a custom build that removes the Okta support and adds in the Duo support, it's easy to do.

Another possibility that has been floated would be to have a configuration parameter for the Okta backend that explicitly tells it to ignore Okta MFA status. #3908 floated something like this but predicated on the idea that in order to set that Duo MFA had to be configured. That's fragile (it could be _unconfigured_ after the bypass had been set), but the other part of that issue is that if we ignore MFA_REQUIRED then we won't actually know whether or not the final state is PASSWORD_EXPIRED. So that opens back up the same exact security issue that 0.9.2/3 were patching. _Knowingly_, true, so that's a key difference, and maybe that's enough to allow it.

**Comment by shanewarner** [comment]
@jefferai What we'd previously done is overlay an MFArequirement with the default policy on auth/token/lookup* so that after Okta authentication succeeded, the MFA was immediately kicked off in order to be able to retrieve the token. We are using the enterprise product that has Okta MFA, though.

This does not work for us at all now and I believe it to be the same issue although you are using a different MFA, it's the same workflow and problem.

**Comment by jefferai** [comment]
@shanewarner Only the CLI actually calls lookup-token on a token in order to show the user some information. Anyone authenticating via the HTTP API (or other CLIs, or other libraries) could simply not call a lookup-self on the token and totally bypass your MFA.

This issue is only about the legacy Duo MFA. I'm not sure what issue you are describing with respect to Okta MFA in Vault Enterprise, but if you are having an issue, you should reach out to enterprise support.

**Comment by shanewarner** [comment]
@jefferai Confirmed the API bypass, thanks. 

In regards to the issue, it's not the MFA method failing, it doesn't even get that far - it's simply okta auth failing. We do have a support ticket open but are still working through it with them. 

To demonstrate that this is an auth failure with okta requiring MFA for a user, I've patched a local copy of 0.9.3 with another case statement including "MFA_REQUIRED" that does nothing like "SUCCESS" and auth works.

**Comment by shanewarner** [comment]
```
       case "SUCCESS":
                // Do nothing here

        case "MFA_REQUIRED": <---- Added this to builtin/credential/okta/backend.go
                // Do nothing here
```
```
vault# bin/vault -version
Vault v0.9.3 ('d285fdf957e3202dfd3f5150365cc3ef62a2b692+CHANGES')
```
```
vault# bin/vault login -method=okta username=shanew
Password (will be hidden):
Success! You are now authenticated.
```

**Comment by jefferai** [comment]
Right, so the issue is that your Okta account requires Okta MFA and the backend does not currently support Okta MFA. Not allowing authentication is the correct behavior up until one of the possibilities mentioned above gets coded. I would also encourage you, in parallel, to contact Okta and ask them to perform MFA in band during login attempts.

**Comment by shanewarner** [comment]
No, that is not the issue.

The enterprise backend supports Okta MFA.  Okta authentication itself has never triggered MFA, even on the Enterprise version, at least when paired with the Okta MFA functionality. The API has always returned MFA_REQUIRED and worked just fine on previous versions.

MFA is then triggered when accessing something that has an MFA requirement in a policy, not during the authentication process.

**Comment by jefferai** [comment]
The issue is that in previous versions Vault was incorrectly ignoring a range of statuses coming from the Okta API, allowing authentication when it should not have.

**Comment by shanewarner** [comment]
Sure that is one issue, but this one is an exception to that.

MFA_REQUIRED is an indication of a successful authentication of the first factor, The MFA factor is completely separate from authentication, and is enforced when a policy is matched like I said before. 

This should be allowed to pass through in order to facilitate functionality of the current workflow of the enterprise product, as it always was before. 

This worked:
vault login -method=okta username=xxxx -> vault read secret/foo (requiring mfa)-> mfa triggers and yields the secrets upon successful mfa

Now it doesn't.

What needs to happen is that the entire implementation be gutted and rewritten properly to implement this so that the associated MFA method is triggered upon receiving MFA_REQUIRED or whatever. The way it's setup now makes that not possible and is a breaking change for paying customers like us.

The real issue here is that stuff like this is making it into enterprise products without being properly vetted and tested first....

**Comment by jefferai** [comment]
> This should be allowed to pass through in order to facilitate functionality of the current workflow of the enterprise product, as it always was before.

It wasn't passed through before; it was simply silently ignored. MFA_REQUIRED means that the username and password were correct, but the administrator has decided that in addition to the username and password, an MFA check must pass. The fact that it worked before was because Vault was not actually doing the proper thing and was ignoring the administrator's specified requirement that an MFA check pass. It was a bug, and a security bug at that.

The change in behavior was clearly called out in the Changelog: only SUCCESS and PASSWORD_WARN are allowed at this time. MFA_REQUIRED was not a part of that list. If you are happy with the previous behavior, or if that change is unacceptable to you, you can simply stay on 0.9.1 until we have a workaround in 0.9.4 that either allows for the Okta MFA flow or for an administrator to _knowingly_ bypass the Okta MFA check, or both. At this point, you at least are aware that staying on 0.9.1 means that status checks are not properly honored.

**Comment by jgiles** [comment]
@jefferai @broamski @shanewarner #3943 is a patch for properly handling Okta-required MFA in the okta auth backend.

I've tested it with our Okta setup (Okta auth requires Duo MFA), and it should still support Okta Push like #3908 but I haven't been able to test that. @shanewarner I'd appreciate if you also tested out the change.

I'd like to add some acceptance tests... but MFA is pretty hostile to automated testing, I'm not sure how to do it.

**Comment by jefferai** [comment]
I've added #3944 which allows the administrator to opt-in to the previous behavior with a simple configuration change. I don't currently have an Okta environment to test with, so any interested parties that can test and report back would be appreciated.

**Comment by jgiles** [comment]
@jefferai I expect either #3943 or #3944 should be merged, and not both?

**Comment by jefferai** [comment]
@jgiles I haven't looked at #3943 yet. I think there are two separate concerns which are complementary, not orthogonal:

* Allowing bypass of Okta MFA requirement (but with knowledge of and action by the administrator)
* Implementing the Okta MFA flow so that it can be done in-band and proper status communicated at the end

#3944 satisfies the first requirement only, but does so in a very simple and straightforward way. It would be sufficient for closing this particular ticket, but a PR satisfying the second requirement would still be great.

**Comment by shanewarner** [comment]
Hi Guys,

Just reporting back on 3943 that we built and tried. This is basically what we're seeing on a fresh vault cluster with okta auth enabled:

# vault login -method okta username=xxxx
Password (will be hidden):
Error authenticating: Error making API request.

URL: PUT https://vaultxxxxx.com/v1/auth/okta/login/xxxxxx
Code: 400. Errors:

* Okta auth failed: 1 error(s) decoding:

* '[1].Profile[keys]' expected type 'string', got unconvertible type '[]interface {}'

**Comment by jgiles** [comment]
@shanewarner  Thank you for testing it out! I think that's a silly mistake that  https://github.com/hashicorp/vault/pull/3943/commits/32f18784d265be79a805df78810d972862879e2e should fix.

**Comment by shanewarner** [comment]
@jgiles 

```
# vault login -method=okta username=xxxx
Password (will be hidden):
Error authenticating: Put https://vaultxxxxx.com/v1/auth/okta/login/xxxx: stream error: stream ID 1; INTERNAL_ERROR
```

We're seeing this, but the okta verify push definitely comes to my phone. It just errors out before I get a chance to approve it. Looks like it's close to working though

**Comment by jefferai** [comment]
@shanewarner please test #3944

**Comment by jgiles** [comment]
@jefferai I've tested #3944 and can confirm it works as intended: With bypass_okta_mfa set, Okta auth completes without any MFA. If I add an mfa_config to the okta backend, I get a Duo push.

@shanewarner to confirm are you using Okta Verify Push or Duo Push for your MFA on Okta login? I'll see if I can get an Okta Verify setup to test with, I've actually only tested with Duo.

I'll investigate that error, but it's a bit trickier to diagnose at a glance - can you reproduce it with debug logging enabled and post the logs? Run `export VAULT_LOG_LEVEL=debug` in the terminal where you launch Vault to enable debug logging.

**Comment by jgiles** [comment]
@shanewarner https://github.com/hashicorp/vault/pull/3943/commits/87ecefaabd46e13e2a3f9ab31b817aa8b0a862f0 fixes the Okta Verify issue in #3943.

It was just another error where we were a little to strict about the JSON we accepted from Okta... they really went for it on the whole HAL thing, it's pretty annoying from a statically typed language.

**Comment by jgiles** [comment]
@jefferai do you want a new issue tracking the work in #3943? Or is the PR on its own enough?

I've successfully tested both the Duo Push and the Okta Verify Push integrations.
